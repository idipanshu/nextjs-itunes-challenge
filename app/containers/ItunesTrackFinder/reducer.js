import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: searchContainerTypes, Creators: searchContainerCreators } = createActions({
  requestGetSongs: ['searchedTerm'],
  requestGetSongDetails: ['trackId'],
  successGetSongs: ['data'],
  successGetSongDetails: ['data'],
  failureGetSongs: ['error'],
  clearSongs: []
});

export const initialState = { searchedTerm: null, trackId: null, songsData: {}, trackDetails: {}, songsError: null };

/* eslint-disable default-case, no-param-reassign */
export const searchContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case searchContainerTypes.REQUEST_GET_SONGS:
        draft.searchedTerm = action.searchedTerm;
        break;

      case searchContainerTypes.CLEAR_SONGS:
        draft.searchedTerm = null;
        draft.songsData = {};
        draft.songsError = null;
        break;

      case searchContainerTypes.SUCCESS_GET_SONGS:
        draft.songsData = action.data;
        break;

      case searchContainerTypes.REQUEST_GET_SONG_DETAILS:
        draft.trackId = action.trackId;
        break;

      case searchContainerTypes.CLEAR_SONG_DETAILS:
        draft.trackId = null;
        draft.trackDetails = {};
        draft.songsError = null;
        break;

      case searchContainerTypes.SUCCESS_GET_SONG_DETAILS:
        draft.trackDetails = action.data;
        break;

      case searchContainerTypes.FAILURE_GET_SONGS:
        draft.songsError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default searchContainerReducer;
