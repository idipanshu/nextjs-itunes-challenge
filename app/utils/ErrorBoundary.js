import React, { Component } from 'react';

export default class ErrorBoundary extends Component {
  constructor(props) {
    super(props);

    this.state = {
      error: false
    };
  }

  static getDerivedStateFromError(error) {
    if (error) {
      return { error: true };
    }
  }

  componentDidCatch(error) {
    throw Error(error);
  }

  render() {
    if (this.state.error) {
      return (
        <div>
          <h1>Something went wrong.</h1>
        </div>
      );
    }

    return this.props.children;
  }
}
