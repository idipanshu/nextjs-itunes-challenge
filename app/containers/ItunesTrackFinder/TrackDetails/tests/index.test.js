import React from 'react';
import { renderProvider, timeout } from '@app/utils/testUtils';

import { searchContainerTypes } from '../../reducer';
import { SongDetailsPageTest as SongDetailsPage, mapDispatchToProps } from '../index';

describe('<TrackDetails /> tests', () => {
  let submitSpy;
  const query = { trackId: '11111111111' };

  const mockData = {
    trackId: 741296001,
    artistName: 'Francis Lawrence',
    collectionName: 'Sam Claflin: 4-Film Collection',
    trackName: 'The Hunger Games: Catching Fire'
  };

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(
      <SongDetailsPage dispatchGetSongDetails={submitSpy} fetchedTracks={mockData} match={{ query }} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchGetItunesTracks when no data is present in the store', async () => {
    const searchedTerm = '11111111111';

    renderProvider(<SongDetailsPage dispatchGetSongDetails={submitSpy} fetchedTracks={{}} match={{ query }} />);

    await timeout(500);

    expect(submitSpy).toBeCalledWith(searchedTerm);
  });

  it('should NOT dispatchGetItunesTracks if matching data is found', async () => {
    const data = { 11111111111: {} };

    renderProvider(
      <SongDetailsPage
        songsData={{ ...data }}
        dispatchGetSongDetails={submitSpy}
        fetchedTracks={{}}
        match={{ query }}
      />
    );

    await timeout(500);

    expect(submitSpy).toBeCalledTimes(0);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchSongsDataSpy = jest.fn();
    const trackId = '11111111111';

    const actions = {
      dispatchGetSongDetails: { trackId, type: searchContainerTypes.REQUEST_GET_SONG_DETAILS }
    };

    const props = mapDispatchToProps(dispatchSongsDataSpy);
    props.dispatchGetSongDetails(trackId);
    expect(dispatchSongsDataSpy).toHaveBeenCalledWith(actions.dispatchGetSongDetails);
  });
});
