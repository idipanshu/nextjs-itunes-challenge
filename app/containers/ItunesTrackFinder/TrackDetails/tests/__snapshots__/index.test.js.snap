// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackDetails /> tests should render and match the snapshot 1`] = `
.c5 {
  white-space: pre-line;
}

.c15 {
  max-width: 100%;
  margin: auto;
  border: 1px solid red;
  margin: 0.5rem auto;
  border-radius: 10px;
  background-color: #f5f5f5;
}

.c7.c7 {
  background-color: #f7f0e9;
  border: none;
  border-radius: 10px;
  box-shadow: 0 0 10px rgba(2,2,2,0.4);
  cursor: pointer;
  -webkit-transition: -webkit-transform 300ms ease-in-out;
  -webkit-transition: transform 300ms ease-in-out;
  transition: transform 300ms ease-in-out;
}

.c7:hover {
  background-color: #b59862;
  -webkit-transform: scale(1.025);
  -ms-transform: scale(1.025);
  transform: scale(1.025);
}

.c8 {
  font-size: 1.2rem;
  color: #020826;
}

.c13 {
  margin: 0.7rem 0;
  text-align: justified;
  font-family: monospace;
  font-weight: 400;
}

.c14 {
  font-weight: bold;
}

.c12 {
  font-family: Monospace;
  font-size: 0.9em;
  padding: 0.2rem 0.4rem;
  color: #f8f8f8;
  background-color: #111;
  border-radius: 5px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 0.6rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 175px;
  overflow: hidden;
}

.c11 {
  color: #020826;
  padding-inline: 0.2rem;
}

.c10 {
  border-radius: 5%;
  width: 150px;
  height: 150px;
}

.c0 {
  padding: 1rem;
  border: 1px solid #111;
  border-radius: 10px;
  max-width: 85%;
  margin: 1rem auto;
  box-shadow: 0 0 5px rgba(2,2,2,0.2);
}

.c1 {
  font-size: 1.7rem;
  font-weight: 600;
  color: #b59862;
  margin: 0.5rem 0;
}

.c4 {
  font-weight: bold;
  font-family: sans-serif;
  margin-top: 0.75rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 1rem;
}

.c3 {
  padding: 1rem;
  width: 30%;
  border: 3px solid #b59862;
  border-radius: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-shadow: 0 0 2px #222;
}

.c6 {
  width: 70%;
}

@media (max-width:1000px) {
  .c2 {
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
  }
}

@media (max-width:1000px) {
  .c3 {
    width: 100%;
  }
}

@media (max-width:1000px) {
  .c6 {
    width: 100%;
  }
}

<body>
  <div>
    <div
      class="c0"
    >
      <a
        href="/"
      >
        Back to Search
      </a>
      <h1
        class="c1"
      >
        Francis Lawrence
      </h1>
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <p
            class="c4"
          >
            Collection:  
          </p>
           
          <span
            class="c5"
            data-testid="text"
          >
            Sam Claflin: 4-Film Collection
          </span>
          <p
            class="c4"
          >
            Track Name:  
          </p>
           
          <span
            class="c5"
            data-testid="text"
          >
            The Hunger Games: Catching Fire
          </span>
        </div>
        <div
          class="c6"
        >
          <div
            class="ant-card ant-card-bordered c7"
            data-testid="music-card"
          >
            <div
              class="ant-card-body"
            >
              <h2
                class="c8"
              >
                The Hunger Games: Catching Fire
              </h2>
              <div
                class="c9"
              >
                <img
                  alt="album art"
                  class="c10"
                />
                <div
                  class="c11"
                >
                  <span
                    class="c12"
                  >
                    Explicit Content
                  </span>
                  <p
                    class="c13"
                  >
                    Francis Lawrence
                  </p>
                  <p
                    class="c13"
                  >
                    Price
                    :
                     
                    <span
                      class="c14"
                    >
                       
                    </span>
                  </p>
                </div>
              </div>
              <div
                class="c15"
              >
                <audio
                  controls=""
                  id="741296001"
                >
                  <source
                    type="audio/mp3"
                  />
                  Your browser does not support audio tags
                </audio>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
